src/ILCompiler.Compiler/src/IL/Stubs/StartupCode/StartupCodeMainMethod.cs

```diff
@@ -145,5 +145,69 @@ public override bool IsNativeCallable
                 return true;
             }
         }
+
+        /// <summary>
+        /// Wraps the main method in a layer of indirection. This is necessary to protect the startup code
+        /// infrastructure from situation when the owning type of the main method cannot be loaded, and codegen
+        /// is instructed to generate a throwing body. Without wrapping, this behavior would result in
+        /// replacing the entire startup code sequence with a throwing body, causing us to enter managed
+        /// environment without it being fully initialized. (In particular, the unhandled exception experience
+        /// won't be initialized, making this difficult to diagnose.)
+        /// </summary>
+        private class MainMethodWrapper : ILStubMethod
+        {
+            private MethodDesc _mainMethod;
+
+            public MainMethodWrapper(TypeDesc owningType, MethodDesc mainMethod)
+            {
+                _mainMethod = mainMethod;
+
+                OwningType = owningType;
+            }
+
+            public override TypeSystemContext Context
+            {
+                get
+                {
+                    return OwningType.Context;
+                }
+            }
+
+            public override TypeDesc OwningType
+            {
+                get;
+            }
+
+            public override string Name
+            {
+                get
+                {
+                    return "StartupCodeSafetyWrapper";
```

Comment:
Nit: I would call it `MainMethodWrapper`. Safety and similar always get people confused because of safety/security has many angles.